packages/next/src/server/api-utils/node.ts

```diff
@@ -530,23 +530,31 @@ export async function apiResolver(
     }
 
     // Call API route method
-    await getTracer().trace(
+    const apiRouteResult = await getTracer().trace(
       NodeSpan.runHandler,
       {
         spanName: `executing api route (pages) ${page}`,
       },
       () => resolver(req, res)
     )
 
-    if (
-      process.env.NODE_ENV !== 'production' &&
-      !externalResolver &&
-      !isResSent(res) &&
-      !wasPiped
-    ) {
-      console.warn(
-        `API resolved without sending a response for ${req.url}, this may result in stalled requests.`
-      )
+    if (process.env.NODE_ENV !== 'production') {
+      if (typeof apiRouteResult !== 'undefined') {
+        if (apiRouteResult instanceof Response) {
+          throw new Error(
+            'API handler returned a Response object in the Node.js runtime, this is not supported. Please use `runtime: "edge"` instead: https://nextjs.org/docs/api-routes/edge-api-routes'
```

Comment:
```suggestion
            'API route returned a Response object in the Node.js runtime, this is not supported. Please use `runtime: "edge"` instead: https://nextjs.org/docs/api-routes/edge-api-routes'
```

`handler` is the name we use for `app` and does support `Response` 👍